rosbag2_compression/src/rosbag2_compression/sequential_compression_writer.cpp

```diff
@@ -81,7 +81,44 @@ void SequentialCompressionWriter::compression_thread_fn()
         });
 
       if (!compressor_message_queue_.empty()) {
-        message = compressor_message_queue_.front();
+        // since the compression operation will manipulate memory inplace, thus
+        // if there are multiple writers with message compression, different manipulation
+        // on the same memory address will triggered, which cause problem. This problem is
+        // described in issue: https://github.com/ros2/rosbag2/issues/1262
+        auto org_message = compressor_message_queue_.front();
+
+        // copy memory of the original message to prevent memory issues
+        message = std::make_shared<rosbag2_storage::SerializedBagMessage>();
+        // initialize buffer
+        message->serialized_data = std::shared_ptr<rcutils_uint8_array_t>(
+          new rcutils_uint8_array_t,
+          [this](rcutils_uint8_array_t * msg) {
+            auto fini_return = rcutils_uint8_array_fini(msg);
+            delete msg;
+            if (fini_return != RCUTILS_RET_OK) {
+              ROSBAG2_COMPRESSION_LOG_ERROR_STREAM(
+                "Failed to destroy serialized message: " << rcutils_get_error_string().str);
+            }
+          });
+        rcutils_allocator_t allocator = rcutils_get_default_allocator();
+        auto ret = rcutils_uint8_array_init(
+          message->serialized_data.get(),
+          org_message->serialized_data->buffer_capacity,
+          &allocator);
+        if (ret != RCUTILS_RET_OK) {
+          ROSBAG2_COMPRESSION_LOG_ERROR_STREAM(
+            "Failed to initialize memory when copy message: " << rcutils_get_error_string().str);
+        }
+        // copy message buffer content
+        memcpy(
+          message->serialized_data->buffer,
+          org_message->serialized_data->buffer,
+          org_message->serialized_data->buffer_length);
+        message->topic_name = org_message->topic_name;
+        message->time_stamp = org_message->time_stamp;
+        message->serialized_data->buffer_capacity = org_message->serialized_data->buffer_capacity;
+        message->serialized_data->buffer_length = org_message->serialized_data->buffer_length;
+
```

Comment:
@zeal-up Could you please move this deep copy piece of code to the `write` 
 method  https://github.com/ros2/rosbag2/blob/c7b17af1eb23625f9db76371a18d9c83076562ea/rosbag2_compression/src/rosbag2_compression/sequential_compression_writer.cpp#L306-L316
Just before locking `compressor_queue_mutex_`?
It will copy messages out of the mutex lock and will minimize the chances that the inner queue will be overflowed and messages will be dropped.